### PR TITLE
check if data size exceeds java array vm limit when it is copied in jni

### PR DESCRIPTION
**Pull request by azagrebin**
to address issue #3849

### CODE REVIEW

File: java/rocksjni/portal.h [hunk 1]
```diff
@@ -4474,25 +4458,21 @@ class JniUtil {
...
+      static const size_t MAX_JARRAY_SIZE = ((static_cast<size_t>(1)) << 31) - 9;
+      if (size > MAX_JARRAY_SIZE) {
+        rocksdb::RocksDBExceptionJni::ThrowNew(env, "Requested array size exceeds VM limit");
         return nullptr;
       }
-
+      
+      const jsize jlen = static_cast<jsize>(size);
+      jbyteArray jbytes = env->NewByteArray(jlen);
+
```
[the remark this answers is not in the document]
Comment (azagrebin) [reply java/rocksjni/portal.h]: thanks, updated

**Comment by azagrebin** [comment]
not sure why travis fails (Error 137), maybe creating of 2.2Gb value is too big for it or the unit test takes too long

File: java/rocksjni/portal.h [hunk 2]
```diff
@@ -4474,25 +4458,25 @@ class JniUtil {
...
-     *
-     * @param env A pointer to the java environment
-     * @param bytes The bytes to copy
-     *
-     * @return the Java byte[] or nullptr if an exception occurs
-     */
-    static jbyteArray copyBytes(JNIEnv* env, const Slice& bytes) {
-      const jsize jlen = static_cast<jsize>(bytes.size());
-
+    static jbyteArray createJavaByteArrayWithSizeCheck(JNIEnv* env, const char* bytes, const size_t size) {
```
Comment (adamretter) [inline java/rocksjni/portal.h]: Could you add some function documentation please?
In particular can you add an `@throws`.

I have been using the Javadoc style even though it is C++, as there are some C++ documenters that will happily read javadoc style comments from C++.

File: java/rocksjni/portal.h [hunk 3]
```diff
@@ -4474,25 +4458,25 @@ class JniUtil {
...
+      // Limitation for array size is vm specific
+      // Current HotSpot VM limitation for array size is Integer.MAX_VALUE - 5 (2^31 - 1 - 5)
+      // Integer.MAX_VALUE - 8 should be safe enough
+      static const size_t MAX_JARRAY_SIZE = ((static_cast<size_t>(1)) << 31) - 9;
+      if(size > MAX_JARRAY_SIZE) {
+        rocksdb::RocksDBExceptionJni::ThrowNew(env, "Requested array size exceeds VM limit");
```
Comment (adamretter) [inline java/rocksjni/portal.h]: The though occurs to me that, we are now potentially throwing a `RocksDBException` for any caller of `copyBytes` which we were not previously.

So we need to look for each Java function that eventually calls this code and make sure it has an `@throws RocksDBException` on both itself and its native interface function. A quick `grep -r "copyBytes" java/rocksdbjni` shows that we will in the least need changes to `WriteBatch.Handler` functions, amongst probably others.

**Comment by adamretter** [comment]
@azagrebin Regards your Travis problem, it seems you ran out of "resources", see: https://github.com/travis-ci/travis-ci/issues/4665

This document (https://docs.travis-ci.com/user/reference/overview/) seems to suggest that we have 4GB available to the build, but I am not sure how much will be used up by Rocks internally etc.

I think for your JUnit test, we need to add an assumption (https://junit.org/junit4/javadoc/4.12/org/junit/Assume.html) that we have enough memory present to run your test. You could calculate the available memory (https://stackoverflow.com/questions/6220790/whats-the-best-way-to-calculate-available-memory-in-java#16517624) and check you have enough for your array in an Assumption.

We could also add a test in C++ to call the `createJavaByteArrayWithSizeCheck` function with some specific sizes, and check that it sets an exception. We are using Google's GTest framework. Is such a test something you might want to add?

File: java/rocksjni/portal.h (hunk 3 above)
Comment (azagrebin) [reply java/rocksjni/portal.h]: As I understand, `WriteBatch.Handler` methods  are called from JNI code which is behind `WriteBatch.iterate` (marked as throws `RocksDBException`). `copyBytes` is called from there before, to prepare key/value for `WriteBatch.Handler`. `WriteBatch.Handler` methods  are user specific java code.

I also found couple of other places which can reach `copyBytes`.

**Comment by azagrebin** [comment]
@adamretter
the JNI code seems to have only java tests, as I understand the c++ tests are setup only for main db code (in make etc).

There is an option to expose test JNI method which calls `createJavaByteArrayWithSizeCheck` and test Exception from Java but not sure about this idea.

**Comment by azagrebin** [comment]
@adamretter 
do you think java test is enough after last changes?
or we need to setup GTest for jni c++ code or there is a better way to add it?

**Comment by azagrebin** [comment]
@adamretter 
can this PR be merged or C++ test or something else still needs to be added?

**Comment by adamretter** [comment]
@azagrebin yes I think it should be okay to merge. @sagar0 / @gfosco Can you merge please?

**Pull request closed without merging**